tfhe/src/shortint/engine/mod.rs

```diff
@@ -239,18 +242,25 @@ impl ShortintEngine {
     fn generate_accumulator_bivariate_with_engine<F>(
         server_key: &ServerKey,
         f: F,
-    ) -> EngineResult<Accumulator>
+        factor: MessageModulus,
+    ) -> EngineResult<BivariateAccumulator>
     where
         F: Fn(u64, u64) -> u64,
     {
-        let modulus = server_key.message_modulus.0 as u64;
+        let factor_u64 = factor.0 as u64;
+        let message_modulus = server_key.message_modulus.0 as u64;
         let wrapped_f = |input: u64| -> u64 {
-            let lhs = (input / modulus) % modulus;
-            let rhs = input % modulus;
+            let lhs = (input / factor_u64) % message_modulus;
+            let rhs = (input % factor_u64) % message_modulus;
```

Comment:
I feel like this may not work if factor_u64 > message modulus ? As the modulo message_modulus is going to remove some data from the input % factor_u64 or like in the original code maybe it's not needed ?

Does it mean we need some asserts ?

Reply:
No it is needed:

Here's me trying to hopefully explain it without being wrong 🤔 

# base

For bivariate pbs, we pack the messages into one with the formula

`v = (lhs * f) + rhs` 
with:
  * ` v < (Carry Modulus * Message Modulus),`
 *  ` rhs < f`


To get back our lhs and rhs from v we do:

lhs:
`v / f = ((lhs * f) / f) + (rhs / f)`
since we have `rhs < f` and we are on unsigned integers: `rhs /f = 0`
`v / f = (lhs * f / f) + (0)`
`v / f = lhs`


rhs:

`v % f = ((lhs * f) + rhs) % f`
`v % f = rhs`

---

Lets say we are doing a call to functionnal bivariate bps,
so we give bivariate function, and in 'one go' it will internally 
generate a LUT and apply it via PBS.

And we are using some param_message_x_carry_y with y >= x.


In that scenario, the factor f that is going to be used is `rhs.degree + 1`

The precondition (is_functional_bivariate_pbs_possible) is:
`(lhs.degree * (rhs.degree + 1)) + rhs.degree < 2^(x+y)`
i.e: can we do `v = (lhs * (rhs.degree + 1)) * rhs` without overflowing and losing information

## Case: the precondition is true:

Then here we don't know if `rhs.degree + 1` is inferior or superior to 2^x the message modulus
ie are their non zero carry bits in either lhs or rhs (or both potentially) ? 

If rhs has some carry bits our factor `f` (`rhs.degree + 1`) we have `f > 2^x` the message modulus

meaning that when doing `v % f = rhs` we have `0 <= rhs < f`
but since `f > message_modulus` when applying the users function we need to apply the message modulus
(ie clean carries) otherwise this will lead to incorrect result:

## Example:

Lets consider some clean and non clean ciphertexts as input for a bivariate pbs:

Pamameters: param_message_1_carry_3,
lhs = 1 % 2^1 = 1 => 0b000|1
rhs = 2 % 2^1 = 0 => 0b001|0

rhs.degree = 3 => f = rhs.degree + 1 = 4

v = (lhs * f) + rhs = (1 * 4) + 2 = 6
v <= 2^4
lhs = v / f = 6 / 4 = 1
rhs = v % f = 6 % 4 = 2

now say the bivariate function is `<` if we dont apply the message_mod
lhs < rhs ?
-> 1 < 2
-> true

if we apply modulo:
(lhs & message_mod) < (rhs % message_mod) ? 
-> (1 * 2) < (2 % 2)
-> 1 < 0
-> false

And since param_message_1_carry_3 means 1 bit of message,
semantically, the message stays on one bit so we have to apply the modulus to
stay correct.

(same idea if lhs has some carry bits)

## Case: the precondition is false

If the precondition is false, we do a clean of the carries (message_extract)
on both of the ciphertext, which will make it so that our 2 ciphertexts will
have a degree of message_mod - 1, thus we can safely do the computations described above
is just that in that case `f = rhs.degree + 1 = message_mod -1 + 1` => `f = message_mod`.


And in fact the whole idea of using `rhs.degree + 1` as the factor `f` instead
of `message_mod` as I think we had in earlier version of concrete-shortint, is that we can avoid
doing some pbses that are just meant to clean some carries if we can afford to clean them durring
the bivariate pbs.

Reply:
Ok, I think I was missing the part where the degree > message_modulus in my reasoning, thanks for the detailed answer